test/parallel_api/numeric/numeric.ops/exclusive_scan_by_segment.pass.cpp

```diff
@@ -16,73 +16,121 @@
 #include "oneapi/dpl/execution"
 #include "oneapi/dpl/algorithm"
 #include "oneapi/dpl/iterator"
+#include "oneapi/dpl/complex"
 
 #include "support/test_config.h"
 #include "support/utils.h"
+#include "support/scan_serial_impl.h"
 
 #if TEST_DPCPP_BACKEND_PRESENT
-#include "support/utils_sycl.h"
+#    include "support/utils_sycl.h"
 
 using namespace oneapi::dpl::execution;
 #endif // TEST_DPCPP_BACKEND_PRESENT
 using namespace TestUtils;
 
-DEFINE_TEST(test_exclusive_scan_by_segment)
+// This macro may be used to analyze source data and test results in test_exclusive_scan_by_segment
+// WARNING: in the case of using this macro debug output is very large.
+// #define DUMP_CHECK_RESULTS
+
+DEFINE_TEST_2(test_exclusive_scan_by_segment, BinaryPredicate, BinaryOperation)
 {
     DEFINE_TEST_CONSTRUCTOR(test_exclusive_scan_by_segment)
 
-    // TODO: replace data generation with random data and update check to compare result to
-    // the result of a serial implementation of the algorithm
-    template <typename Accessor1, typename Accessor2, typename Accessor3, typename Size>
-    void
-    initialize_data(Accessor1 host_keys, Accessor2 host_vals, Accessor3 host_val_res, Size n)
+    template <typename Iterator1, typename Iterator2, typename Iterator3, typename Size>
+    void initialize_data(Iterator1 host_keys, Iterator2 host_vals, Iterator3 host_val_res, Size n)
     {
-        //T keys[n1] = { 1, 2, 3, 4, 1, 1, 2, 2, 3, 3, 4, 4, 1, 1, 1, ...};
-        //T vals[n1] = { 1, 1, 1, ... };
+        //T keys[n] = { 1, 2, 3, 4, 1, 1, 2, 2, 3, 3, 4, 4, 1, 1, 1, ...};
+        //T vals[n] = { n random numbers between 0 and 4 };
 
-        int segment_length = 1;
-        int i = 0;
-        while (i != n)
+        ::std::srand(42);
+        Size segment_length = 1;
+        for (Size i = 0; i != n;)
```

Comment:
I understand that it was done before your changes but these nested loops look really weird

Reply:
I have refactored this piece of code to use a single loop and be a bit more clear.